app/boot_control/grub.py

```diff
@@ -27,48 +40,458 @@
 )
 
 
-class GrubController(
-    VersionControlMixin, OTAStatusMixin, SlotInUseMixin, BootControllerProtocol
-):
-    def __init__(self) -> None:
+@dataclass
+class GrubMenuEntry:
+    """
+    NOTE: should only be called by the get_entry method
+
+    linux: vmlinuz-<ver>
+    linux_ver: <ver>
+    initrd: initrd.img-<ver>
+    """
+
+    linux_ver: str
+    title: str
+    menuentry: str
+    linux: str
+    initrd: str
+    rootfs_uuid: str
+
+    def __init__(self, ma: re.Match) -> None:
+        """
+        NOTE: check GrubHelper for capturing group definition
+        """
+        self.linux_ver = ma.group("kernel_ver")
+        self.title = ma.group("title")
+        self.menuentry = ma.group("menu_entry")
+
+        # get linux and initrd from menuentry
+        if linux_ma := GrubHelper.linux_pa.search(self.menuentry):
+            self.linux = linux_ma.group("kernel")
+            self.rootfs_uuid = linux_ma.group("rootfs_uuid")
+        if initrd_ma := GrubHelper.initrd_pa.search(self.menuentry):
+            self.initrd = initrd_ma.group("initrd")
+
+        assert self.linux and self.initrd, "failed to detect linux img and initrd"
+        assert self.rootfs_uuid, "failed to detect rootfs uuid"
+
+
+class GrubHelper:
+    menuentry_pa: ClassVar[re.Pattern] = re.compile(
+        # whole capture group
+        r"^(?P<menu_entry>\s*menuentry\s+"
+        # menuentry title:
+        # format: <os>, with Linux <kernel_ver>[ [<recovery>] [<os_probe>]]
+        r"'(?P<title>[^,]*, +with +Linux +(?P<kernel_ver>[^'\s]*)(?P<other>( *\([^\)]*\))*)?)'"
+        r"[^\{]*"  # menuentry options
+        r"\{(?P<entry>[^\}]*)\}"  # menuentry block
+        r")",  # end of whole capture
+        re.MULTILINE | re.DOTALL,
+    )
```

Comment:
the submenu is supported?
After installing OS by USB installer, a submenu is used.

Reply:
Does this refactoring is required??

Reply:
Thanks for reporting! I will add submenu support.

Reply:
@obi-t4 
After checking the implementation in this PR and rethink a while, for the new implementation of grub controller, 
extra adjustment is not needed for supporting freshly installed system.
I will do test to confirm whether the above statement is true or not.

Explanation as follow:
In the new implementation of grub controller, when initializing ota-partition files for new installed system
(a system that doesn't have `ota-partition` configured, and also doesn't boot with `ota-partition` mechanism), 
**old `grub.cfg` will NEVER be used**, and new generated `grub.cfg` by otaclient grub controller will ALWAYS set `GRUB_DISABLE_SUBMENU=y`.
So no matter `grub submenu` is enabled or not for the system,  the grub controller will initialize without any problem.

Reply:
Comments for the explanation are added in e958cb1.